### ISSUE TITLE
Refinement list doesn't works with ES5

### ISSUE THREAD

**Issue by nicolapagni** [searchkit/searchkit#319]
Hi,

I tried to use the _Refinement list_ component using it with ES5.
If you try to interact with it, probabily you will receive the following:

> Unknown key for a START_OBJECT

Analyzing the problem, I understand that the main problem is the different mapping between ES2.x and the last majar ES5. In details, until elastis 2.x wen could use the type "String", instead in the new one this type is not available. Now we can use the type "Text".

For this component, in the official guide, is specified that we have to use the type "String", for example:

```
{
  "commercial" : {
    "properties" : {
      "ProductName" : {
        "type" : "string",
        "fields" : {          
          "raw" : {
            "type" : "string",
            "index" : "not_analyzed"
          }
        }
      }
    }
  }
}
```

When searchKit will compatible with the new major of elastic?

Thanks in advance.

Regards, 
  Nicola

**Comment by GregoryPotdevin** [comment]
Unless there are major changes to the query DSL it should work fine.

ElasticSearch 5 isn't out yet so it seems a bit premature to update the documentation for it. 

For reference for those who didn't know : 

## String
https://www.elastic.co/guide/en/elasticsearch/reference/master/string.html

> The string field has been removed in favor of the text and keyword fields.

## Text
https://www.elastic.co/guide/en/elasticsearch/reference/master/text.html

> A field to index full-text values, such as the body of on email or the description of a product. These fields are analyzed, that is they are passed through an analyzer to convert the string into a list of individual terms before being indexed. 

## Keyword
https://www.elastic.co/guide/en/elasticsearch/reference/master/keyword.html

> A field to index structured content such as email addresses, hostnames, status codes, zip codes or tags.
> They are typically used for filtering (Find me all blog posts where status is published), for sorting, and for aggregations. Keyword fields are ony searchable by their exact value.

**Comment by nicolapagni** [comment]
> Unless there are major changes to the query DSL it should work fine.

In the indexing & mapping section of the guide, I applied to my mapping this example:
```

{
  "movie" : {
    "properties" : {
      "genres" : {
        "type" : "string",
        "fields" : {          
          "raw" : {
            "type" : "string",
            "index" : "not_analyzed"
          }
        }
      }
    }
  }
}
```

Using it with ES2 all works fine...
Same test with ES5, when i tried to click on an option of refinement list, it give me the error I wrote in my prev post.

So, How could I change the mapping to make it compatible with ES5?

**Comment by GregoryPotdevin** [comment]
Sorry, maybe I wasn't clear. I don't mean that it should work perfectly following the documentation for the setup. Searchkit doesn't take care of the mapping part and that section in the documentation is pretty much "recommendations". ES5 changes the mapping, and the new syntax they have is to use "text" for fulltext search and "keyword" for facetted entries (those that had "not_analyze"). I haven't tried but something like this might work : 

```json
{
  "movie" : {
    "properties" : {
      "genres" : {
        "type" : "keyword",
      }
    }
  }
}
```

That field doesn't have the "raw" subfield anymore (it's baked in using ES5, I guess), so compared to the documentation you'll want to use just `field="genres"` and not `field="genres.raw"`. If only the mapping changed the rest should work fine, but I haven't checked.

**Comment by nicolapagni** [comment]
I just tried and unfortunately it not seems works.
I used the syntax you gave me for the mapping, and then i removed the .raw from assignment of the field. The exact error of the response I have is the following:

`{"error":{"root_cause":[{"type":"parsing_exception","reason":"Unknown key for a START_OBJECT in [filter].","line":1,"col":11}],"type":"parsing_exception","reason":"Unknown key for a START_OBJECT in [filter].","line":1,"col":11},"status":400}`

Obviously, if I click on the "ALL" label of the filter I don't see any errors.

When you have time to spend on it let me know how are the tests.

Thankssss,
  Nicola

**Comment by GregoryPotdevin** [comment]
Apparently the list of changes for the query DSL is really big, we'll have to see to what extent we can have the current code work for both ES2.X and ES5, or if we need to fork into like Searchkit 2 to support ES5.

## Query DSL changes 
https://www.elastic.co/guide/en/elasticsearch/reference/master/breaking_50_search_changes.html
@ssetem just some random changes : 
> `limit` => Use the terminate_after parameter instead.
> `query` => Is obsolete after filters and queries have been merged.
> Removed support for the deprecated filter fields in nested query, in favour of query.
> Removed support for the deprecated minimum_should_match and disable_coord in terms query, use bool query instead.
> Removed support for the top level filter element in function_score query, replaced by query.
> Removed support for the deprecated top level filter in the search api, replaced by post_filter
> Removed support for multiple highlighter names, the only supported ones are: plain, fvh and postings

## Mapping changes

https://www.elastic.co/guide/en/elasticsearch/reference/master/breaking_50_mapping_changes.html

> string fields replaced by text/keyword fields
> Default string mappings
```json
{
  "type": "text",
  "fields": {
    "keyword": {
      "type": "keyword",
      "ignore_above": 256
    }
  }
}
```
(if I understand this correctly, it means that by default we don't need the raw subfields anymore, all the fields will have them. So in the queries we just replace "field.raw" by "field.keyword")

@nicolapagni we'll get back at you as soon as we have more news. Do you know when ES5 is supposed to be released ?

**Comment by nicolapagni** [comment]
> (if I understand this correctly, it means that by default we don't need the raw subfields anymore, all the fields will have them. So in the queries we just replace "field.raw" by "field.keyword")

Through kibana/console i tried again with keyword instead of raw. In this way I don't see errors and the response looks like as expected.
Using the same query in searchkit I receive the same error. 
That is strange because the response format doesn't changes...

@GregoryPotdevin Actualy we are at alpha-4. I can suppose It will not be released until 2-3 months

**Comment by ssetem** [comment]
yeh we need to investigate how we would support elasticsearch 5.

It would be great if those beginning to use alpha of elastic 5 would helps us in showing the query changes, perhaps even try to get the accessors working etc.

**Comment by joemcelroy** [comment]
this issue should be fixed in #392
